docs/design/bin_pack_fault_domains.md

```diff
@@ -0,0 +1,126 @@
+# Support bin packing of fault domains
+
+## Metadata
+
+* Authors: @johscheuer
+* Created: 2021-10-01
+* Updated: 2021-10-01
+
+## Background
+
+In the current implementation of the operator we try to distribute our Pod across as many fault domains as possible.
+If a cluster has more fault domains than the operator creates Pods for a specific class all Pods will be on a different fault domain.
+This has the drawback that the operator can only interact on a single Pod for safe operations e.g. to minimize the risk of data loss.
+For clusters with a non-trivial amount of Pods this means that some operational tasks will take very long.
+A better way would be to allow the operator to bin pack Pods in the same fault domain.
+
+## General Design Goals
+
+* Provide a way how a user can specify the number of logical fault domains
+* The user should be able to decide if the number of logical fault domains is a requirement (default should be optional)
+
+## Current Implementation
+
+The current implementation adds a `preferredDuringSchedulingIgnoredDuringExecution` `PodAntiAffinity` to every Pod [code](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/internal/pod_models.go#L308-L334).
+The label selector will contain the process class and the cluster name, to ensure processes with the same class are distributed across different fault domains is possible.
+A user can define additional `PodAntiAffinity` but can't prevent the operator from createn the default `PodAntiAffinity`.
+The only expection is the specival fault domain `foundationdb.org/none` and `foundationdb.org/kubernetes-cluster`.
+
+## Proposed Design
+
+The idea would be to use a custom label like `foundationdb.org/distribution-key`.
+As value we would use the process number modulo the `DesiredFaultDomains` and a prefix for the `BinPack` mode.
+In addition to that we have to change the `PodAntiAffinity` rule.
+The following would be an example of the `affinity` term for a Pod.
+The example assumes that process number modulo `DesiredFaultDomains` is `0` and we have set `BinPackAll`:
+
+```yaml
+  affinity:
+    # Affinity to schedule Pods with the same distribution key together
+    podAffinity:
+      preferredDuringSchedulingIgnoredDuringExecution:
+      - podAffinityTerm:
+          labelSelector:
+            matchLabels:
+              foundationdb.org/distribution-key: "all-0"
+          topologyKey: fault_domain
+        weight: 1
+    # AntiAffinity to not schedule the Pod on the same fault domain where other Pods with a different distribution key are running
+    podAntiAffinity:
+      preferredDuringSchedulingIgnoredDuringExecution:
```

Comment:
It may be worth calling out that `requiredDuringScheduling` anti-affinity is only allowed on the hostname topology key by default: https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#an-example-of-a-pod-that-uses-pod-affinity.